### ISSUE TITLE
Request: disable weak and suspect crypto

### ISSUE THREAD

**Issue by wolfgangrupprecht** [mkj/dropbear#304]
Crypto flaws and suspect crypto are hard for the normal user to understand or even easily research.   It would be nice if the out of the box defaults were to disable any suspect or weak crypto.   The attached file is the log of "ssh-audit" run against dropbear_2024.85.  

[ssh-audit.log](https://github.com/mkj/dropbear/files/15358512/ssh-audit.log)

**Comment by WolfWings** [comment]
Not a developer on this project, but I try to help issues along in projects I use when I spot things!

So tossing this over the wall without even stripping out the ANSI escape codes which make the file gibberish here on github is not the most approachable to getting this worked efficiently, and this could likely have been better handled as a pull request instead.

This 'audit' is very subjective and pushes for as many restrictions as possible even when there's not a large corpus of evidence for some of the suggestions or when the items in question are still mandatory to the protocol.

- NIST ECDSA curves have no alternative and are widely deployed and used by a lot of systems; their status has been an ongoing debate for decades now with no conclusive evidence they're vulnerable in roughly 20 years.
- diffie-hellman-group14-sha1 has been disabled in OpenSSH since 8.2 (released in early 2020) by default, and they'll be removing it entirely soon, so 100% agreed that should get tossed from the default configuration and likely the whole codebase.
- diffie-hellman-group14-sha256 is mandatory to the SSH protocol on the [latest RFCs](https://www.rfc-editor.org/rfc/rfc9142.html#ffc_named_group) that deprecate diffie-hellman-group14-sha1.
- ssh-rsa should indeed be disabled by default, OpenSSH has since 8.8 released in 2021 because it is indeed broke as a joke.
- hmac-sha1 is mandatory to the SSH protocol, and while sha1 is considered compromised as a hash itself, it's use in an HMAC construct is [still considered valid and safe](https://crypto.stackexchange.com/a/26518) because HMAC does not rely on the collision-resistance of the hash used.
- diffie-hellman-group16-sha512 should be added if/when possible, yes.
- There's no explanation for why adding the 'twofish' ciphers are suggested? Reinforcing that this 'audit' tool is subjective not objective in nature.
- Adding the "GCM" variants of AES is not mentioned, which again more 'why twofish and not this?' subjective vs objective implication.

The above should hopefully help whatever dev picks this up at least know what's in the attached audit log.

**Comment by jbruchon** [comment]
I should have the choice to use whatever I want to use. It is not your or anyone else's job to restrict my freedom to choose. If I want to use weak ciphers then who are you to tell me I can't? I actually have reasons to use certain weak crypto and nothing some "audit" says will change them. This issue should be closed.

**Comment by jbruchon** [comment]
> This is about providing sane and secure defaults.
No, it's not. It's removing options because you're the nanny of the users. If you want to gatekeep weak crypto behind a parameter that says "it's okay, I know what I'm doing" then that's fine, but "disabling" here means disabling in the source, as in not making available at all. "Default" is not the same as "available."

> Petulant screaming into the void about freedoms won't change the relevance of this issue.
"Petulant" is a logic-free insult: _"unreasonably irritable or ill-tempered."_ Calling me petulant only makes you look like a fool that can't defend what they say. I can call you a dickhead but it won't change the validity of my arguments. Try not being a dickhead next time.

> Also it's not like the older versions will cease to be available, so you're welcome to choose to run older, out-dated software if you need older, out-dated ciphers.
Not if I receive a binary built for an old ARM router. This is the trendy "if you don't like it, build your own everything" dismissal that political lunatics use. Try being more rational next time.

> Some of the items listed are things that need to be added, not removed,
The issue title is "disable weak and suspect crypto." Try understanding how context works next time.

> but some of the removals are warranted at this point from the default configuration if not the codebase entirely.
That is your subjective opinion which also has a sample size of 1. Dropbear is deployed on millions of devices, but assuming there are only about one million, your opinion represents 1/1000000 of the potential user base. Appeals to modernity are not logical.

If you weren't also choosing to be narcissistic, you'd have understood that my response was aimed squarely at the original issue poster and you had absolutely nothing to do with it. Did I @ you? No, I didn't.

**Comment by mkj** [comment]
This discussion is pretty tedious. Here isn't a newspaper comments page.

I'll be disabling the sha1 algos by default. I don't think any other changes are required.

**Comment by M95D** [comment]
I have a PR opened for OpenWrt Dropbear package. I added almost all config options available, including all crypto options.
May I ask those of you who understand crypto better than me, to review the defaults and comments in the Config.in ? (Crypto options are at the end.)

https://github.com/openwrt/openwrt/blob/ad30892b1d74942b034b47334ec6c8294466600d/package/network/services/dropbear/Config.in

Thank you!

**Comment by mkj** [comment]
sha1 is now disabled by default in Dropbear 2025.87
